Order build steps execution

gsmet opened the pull request:
This is a revival of what @ppalaga did here: https://github.com/quarkusio/quarkus/pull/6972 .

It's more or less the same patch. I added two additional constraints that were needed.

ppalaga committed:
Fix #1020 Stable main and static init method generation

Co-Authored-By: Guillaume Smet <guillaume.smet@gmail.com>

gsmet commented on extensions/vertx/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxCoreProcessor.java:
```diff
@@ -236,11 +234,10 @@ CoreVertxBuildItem build(
-        }
+        List<Consumer<VertxOptions>> consumers = vertxOptionsConsumers.stream()
+                .sorted()
+                .map(cbi -> cbi.getConsumer())
+                .toList();
```
This change is only there to avoid modifying the list of `vertxOptionsConsumers`.

dmlloyd reviewed (approved):
It mostly LGTM other than a couple minor comments.

By "LGTM" I mean "I don't see any problems here". I do want to reaffirm that we should discuss some of the broader, fundamental issues in a few weeks.

gsmet committed:
Make sure the native executable is always considered first

The order in which ArtifactResultBuildItem has changed because they are
now ordered depending on the build steps dependencies and not ordered
dependending on the timing of the output.

gsmet commented:
Making progress, down to one actual failure...
